File: zig-mcp-template/src/mcp.zig
const std = @import("std");

/// JSON-RPC 2.0 Request
pub const Request = struct {
    jsonrpc: []const u8 = "2.0",
    id: ?std.json.Value = null,
    method: []const u8,
    params: ?std.json.Value = null,
};

/// JSON-RPC 2.0 Response
pub const Response = struct {
    jsonrpc: []const u8 = "2.0",
    id: ?std.json.Value,
    result: ?std.json.Value = null,
    @"error": ?ErrorObject = null,
};

/// JSON-RPC 2.0 Error Object
pub const ErrorObject = struct {
    code: i32,
    message: []const u8,
    data: ?std.json.Value = null,
};

/// Standard JSON-RPC error codes
pub const ErrorCode = enum(i32) {
    ParseError = -32700,
    InvalidRequest = -32600,
    MethodNotFound = -32601,
    InvalidParams = -32602,
    InternalError = -32603,
};

/// MCP Initialize request parameters
pub const InitializeParams = struct {
    protocolVersion: []const u8,
    capabilities: std.json.Value,
    clientInfo: ?ClientInfo = null,
};

/// Client information
pub const ClientInfo = struct {
    name: []const u8,
    version: ?[]const u8 = null,
};

/// MCP Initialize response result
pub const InitializeResult = struct {
    protocolVersion: []const u8,
    capabilities: ServerCapabilities,
    serverInfo: ServerInfo,
};

/// Server capabilities
pub const ServerCapabilities = struct {
    tools: ?ToolsCapability = null,
};

/// Tools capability
pub const ToolsCapability = struct {
    listChanged: ?bool = null,
};

/// Server information
pub const ServerInfo = struct {
    name: []const u8,
    version: []const u8,
};

/// Tool definition
pub const Tool = struct {
    name: []const u8,
    description: []const u8,
    inputSchema: std.json.Value,
};

/// Tools list response
pub const ToolsList = struct {
    tools: []const Tool,
};

/// Tool call request parameters
pub const CallToolParams = struct {
    name: []const u8,
    arguments: ?std.json.Value = null,
};

/// Tool call response result
pub const CallToolResult = struct {
    content: []const Content,
    isError: ?bool = null,
};

/// Content item in tool response
pub const Content = struct {
    type: []const u8,
    text: []const u8,
};

/// Helper to create an error response
pub fn createErrorResponse(allocator: std.mem.Allocator, id: ?std.json.Value, code: ErrorCode, message: []const u8) !Response {
    _ = allocator;
    return Response{
        .id = id,
        .@"error" = ErrorObject{
            .code = @intFromEnum(code),
            .message = message,
        },
    };
}

/// Helper to create a success response
pub fn createSuccessResponse(id: ?std.json.Value, result: std.json.Value) Response {
    return Response{
        .id = id,
        .result = result,
    };
}

File: cpp20-mcp-template/src/mcp.cpp
#include "mcp.hpp"

namespace mcp {

std::string serializeToJson(const rapidjson::Value& value) {
    rapidjson::StringBuffer buffer;
    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
    value.Accept(writer);
    return buffer.GetString();
}

std::string serializeToJson(const rapidjson::Document& doc) {
    rapidjson::StringBuffer buffer;
    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
    doc.Accept(writer);
    return buffer.GetString();
}

rapidjson::Document createSuccessResponse(
    rapidjson::Document::AllocatorType& allocator,
    const rapidjson::Value& id,
    const rapidjson::Value& result
) {
    rapidjson::Document response;
    response.SetObject();

    response.AddMember("jsonrpc", "2.0", allocator);
    response.AddMember("id", rapidjson::Value(id, allocator), allocator);
    response.AddMember("result", rapidjson::Value(result, allocator), allocator);

    return response;
}

rapidjson::Document createErrorResponse(
    rapidjson::Document::AllocatorType& allocator,
    const rapidjson::Value& id,
    ErrorCode code,
    const std::string& message
) {
    rapidjson::Document response;
    response.SetObject();

    response.AddMember("jsonrpc", "2.0", allocator);
    response.AddMember("id", rapidjson::Value(id, allocator), allocator);

    rapidjson::Value errorObj(rapidjson::kObjectType);
    errorObj.AddMember("code", static_cast<int>(code), allocator);
    errorObj.AddMember("message", rapidjson::Value(message.c_str(), allocator), allocator);

    response.AddMember("error", errorObj, allocator);

    return response;
}

rapidjson::Value createInitializeResult(
    rapidjson::Document::AllocatorType& allocator,
    const InitializeResult& result
) {
    rapidjson::Value resultObj(rapidjson::kObjectType);

    resultObj.AddMember("protocolVersion",
        rapidjson::Value(result.protocolVersion.c_str(), allocator), allocator);

    rapidjson::Value capabilities(rapidjson::kObjectType);
    rapidjson::Value tools(rapidjson::kObjectType);
    capabilities.AddMember("tools", tools, allocator);
    resultObj.AddMember("capabilities", capabilities, allocator);

    rapidjson::Value serverInfo(rapidjson::kObjectType);
    serverInfo.AddMember("name",
        rapidjson::Value(result.serverInfo.name.c_str(), allocator), allocator);
    serverInfo.AddMember("version",
        rapidjson::Value(result.serverInfo.version.c_str(), allocator), allocator);
    resultObj.AddMember("serverInfo", serverInfo, allocator);

    return resultObj;
}

rapidjson::Value createToolsList(
    rapidjson::Document::AllocatorType& allocator,
    const std::vector<Tool>& tools
) {
    rapidjson::Value result(rapidjson::kObjectType);
    rapidjson::Value toolsArray(rapidjson::kArrayType);

    for (const auto& tool : tools) {
        rapidjson::Value toolObj(rapidjson::kObjectType);
        toolObj.AddMember("name",
            rapidjson::Value(tool.name.c_str(), allocator), allocator);
        toolObj.AddMember("description",
            rapidjson::Value(tool.description.c_str(), allocator), allocator);
        toolObj.AddMember("inputSchema",
            rapidjson::Value(tool.inputSchema, allocator), allocator);

        toolsArray.PushBack(toolObj, allocator);
    }

    result.AddMember("tools", toolsArray, allocator);
    return result;
}

rapidjson::Value createToolCallResult(
    rapidjson::Document::AllocatorType& allocator,
    const CallToolResult& result
) {
    rapidjson::Value resultObj(rapidjson::kObjectType);
    rapidjson::Value contentArray(rapidjson::kArrayType);

    for (const auto& item : result.content) {
        rapidjson::Value contentObj(rapidjson::kObjectType);
        contentObj.AddMember("type",
            rapidjson::Value(item.type.c_str(), allocator), allocator);
        contentObj.AddMember("text",
            rapidjson::Value(item.text.c_str(), allocator), allocator);
        contentArray.PushBack(contentObj, allocator);
    }

    resultObj.AddMember("content", contentArray, allocator);
    resultObj.AddMember("isError", result.isError, allocator);

    return resultObj;
}

} // namespace mcp
